helix-view/src/editor.rs

```diff
@@ -398,18 +412,88 @@ impl Editor {
-                tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        let doc = &self.documents[&doc_id];
+
+        if !force && doc.is_modified() {
+            anyhow::bail!(
+                "buffer {:?} is modified",
+                doc.relative_path()
+                    .map(|path| path.to_string_lossy().to_string())
+                    .unwrap_or_else(|| "[scratch]".into())
+            );
+        }
+
+        if let Some(language_server) = doc.language_server() {
+            tokio::spawn(language_server.text_document_did_close(doc.identifier()));
+        }
+
+        let mut v = vec![];
+
+        for (view, _focused) in self.tree.views_mut() {
+            view.jumps.remove(&doc_id);
+
+            if let Some(alt) = view.last_accessed_doc {
+                if alt == doc_id {
+                    view.last_accessed_doc = None;
+                }
+            }
+
+            if view.doc == doc_id {
+                let new_id = if let Some(alt) = view.last_accessed_doc {
+                    Some(alt)
+                } else if self.documents.len() > 1 {
+                    // FIXME: We don't store enough "alternate buffer" context -- opening three
+                    // buffers, and then closing two will leave the user with the first buffer still
+                    // there, but not currently visible. To work around this, we use a "best guess"
+                    // -- if there is a buffer that is before it in the tree, switch to that one.
+                    // Otherwise, find the first not-to-be-removed document and switch to that.
+                    self.documents.range(..doc_id).next_back().map_or(
+                        self.documents.iter().find_map(|(&other_id, _)| {
+                            if other_id == doc_id {
+                                None
+                            } else {
+                                Some(other_id)
+                            }
+                        }),
+                        |(&other_id, _)| Some(other_id),
+                    )
+                } else {
+                    None
+                };
+
+                v.push((view.id, new_id, view.doc == doc_id));
             }
-            self.documents.remove(&view.doc);
         }
 
-        self.tree.remove(id);
-        self._refresh();
+        for (view_id, doc_id, doc_is_focused) in v {
+            if doc_is_focused && self.tree.views().count() > 1 {
+                self.close(view_id);
+                continue;
+            }
+
+            let doc_id = if let Some(doc_id) = doc_id {
+                doc_id
+            } else {
+                self.new_document(Document::default())
+            };
+            self.replace_document_in_view(view_id, doc_id);
+            self._refresh();
+        }
+
+        if !self.tree.views().any(|(view, _)| view.doc == doc_id) {
+            self.documents.remove(&doc_id);
```

Comment:
By this point it should be safe to remove the document because you've either removed it from all the views, or you bailed early because the doc was modified.

If you want to keep a check just in case, maybe turn it into a debug_assert?

Reply:
```rust
debug_assert!(!self.tree.views().any(|(view, _)| view.doc == doc_id));
self.documents.remove(&doc_id);
```